src/ot_orchestration/operators/gce.py

```diff
@@ -0,0 +1,733 @@
+        If set as a string, the account must grant the originating account
+        the Service Account Token Creator IAM role.
+        If set as a sequence, the identities from the list must grant
+        Service Account Token Creator IAM role to the directly preceding identity, with first
+        account from the list granting this role to the originating account.
+    """
+
+    def __init__(
+        self,
+        api_version: str = "v2",
+        gcp_conn_id: str = "google_cloud_default",
+        impersonation_chain: str | Sequence[str] | None = None,
+    ) -> None:
+        self._client: LoggingServiceV2AsyncClient | None = None
+        super().__init__(
+            gcp_conn_id=gcp_conn_id,
+            impersonation_chain=impersonation_chain,
+        )
+        self.api_version = api_version
+        self.request_interval = LOGGING_REQUEST_INTERVAL
+
+    def get_conn(self) -> LoggingServiceV2AsyncClient:
+        """Return the Google Cloud Logging service client."""
+        if self._client is None:
+            self._client = LoggingServiceV2AsyncClient(
+                credentials=self.get_credentials(),
+                client_info=CLIENT_INFO,
+            )
+        return self._client
+
+    async def get_exit_code(
+        self,
+        project_name: str,
+        instance_name: str,
+        initial_timestamp: datetime.datetime,
+    ) -> int:
+        """Get the exit code of the startup script of a Google Compute Engine instance.
+
+        According to Google Cloud documentation in `viewing the output of a Linux startup script
+        <https://cloud.google.com/compute/docs/instances/startup-scripts/linux#viewing-output>`__,
+
+        The exit code of the startup script should be visible in serial port 1 or the system logs,
+        but it is not there. The only way to get the exit code is to search the logs for the message
+        by using the logging API.
+
+        The exit code is the number that follows the string "exit status" in the message. To harden
+        the query, we are using the regex:
+
+        startup-script[\w\":\s]*exit status [0-9]+
+
+        This is because the message changes depending on the script exiting successfully:
+
+        startup-script exit status 0
+
+        And when the script fails:
+
+        Script "startup-script" failed with error: exit status 1
+        """  # noqa: D301
+        client = self.get_conn()
+        timestamp_str = initial_timestamp.isoformat()
+        query = f'resource.type="gce_instance" labels.instance_name="{instance_name}" timestamp>"{timestamp_str}" jsonPayload.message=~"startup-script[\w\\\":\s]*exit status [0-9]+"'  # fmt: skip
+        log_pages = None
+
+        while True:
+            try:
+                log_pages = await client.list_log_entries(
+                    resource_names=[f"projects/{project_name}"],
+                    filter=query,
+                    timeout=300,
+                )
+                break
```

Comment:
Keep in mind the try/catch is there mostly because of the cloud logging quota limit, which is 60 requests per minute per service account. If I sleep after every request, instead of taking the amount of time per request (usually milliseconds) * number of requests, this will take the amount of time per request + the interval * number of requests. That can be a whole lot more. :)

My idea was to start sleeping only when we reach that limit, which does not happen often but once it happens will repeat often unless we stop requesting for a while (hence the exponential backoff).

Reply:
Again, I agree, but I would still reset the interval after successful attempt.